Support async signing for V2 channel establishment

jkczyz opened the pull request:
When handling a `tx_complete` message, allow signers to return an error indicating that the signer has not yet complete. This will leave the `ChannelPhase` in an unfunded variant until the signer becomes unblocked. The user calls `ChannelManager::signer_unblocked` to indicate that signing is complete, which will attempt to finish handling the `tx_complete` message again.

Based on #3137.
Fixes #3404.

jkczyz committed:
Support async signing for V2 channel establishment

When handling a tx_complete message, allow signers to return an error
indicating that the signer has not yet complete. This will leave the
ChannelPhase in an unfunded variant until the signer becomes unblocked.

jkczyz committed:
Expose signing session from InteractivelyFunded

To support async signing, the InteractiveTxSigningSession returned when
handling tx_complete needs to be saved in order to retry the
funding_tx_constructed method when the signer is unblocked. This
unfortunately means an Option as needed since the unfunded V2 channel
phase variant is left unfunded until the signer completes.

jkczyz committed:
Finish handling tx_complete in signer_unblocked

Expand ChannelManager::signer_unblocked to finish handling tx_complete
messages. This completes support for async signing in V2 channel
establishment.

jkczyz committed:
Remove Result from into_channel methods

These methods always return Ok, so there is no need to use a Result.

jkczyz committed:
Test async signing in V2 channel establishment

dunxen reviewed (commented):
Seems good to me, but I'm not 100% familiar with the async signing stuff. Seems mostly straightforward though.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9562,6 +9599,45 @@ where
 			self.finish_close_channel(shutdown_result);
 		}
+
+		// Finish any tx_complete handling waiting on async signing.
+		//
+		// TODO: Move this into the earlier channel iteration to avoid duplication and the Vec
+		// allocation once ChannelPhase is refactored into Channel. This can't be avoided with the
+		// current data model because tx_complete handling requires removing the entry from the
+		// channel_by_id map and re-inserting it, which can't be done while iterating over the map.
```
Yup, that's what the comment is referring to.

dunxen replied:
Oh i missed a whole line when reading. Stated the exact same thing 🤦‍♂️

jkczyz committed:
Remove unnecessary clone and collect

jkczyz committed:
Remove unused errs Vec when claiming payments

The errs Vec in ChannelManager::claim_payment_internal hasn't been
populated since commit fea63930831fb5c7ea4c38a2d405d4bdb5ae6f29. Drop it
along with the code that consumed it.

jkczyz committed:
Remove unnecessary drains from channelmanager.rs

jkczyz commented:
Fixed a couple `rustfmt` issues in `dual_funding_tests.rs`. Also, removed a ton of unnecessary `drain`s from `channelmanager.rs` for good measure since I was looking at related code and noticed a `Vec` wasn't being used, which I removed in another commit, too.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -1682,6 +1682,10 @@ pub(super) trait InteractivelyFunded<SP: Deref> where SP::Target: SignerProvider
 
+	fn interactive_tx_signing_session_mut(&mut self) -> &mut Option<InteractiveTxSigningSession>;
+
+	fn interactive_tx_signing_session_with_context_mut(&mut self) -> (&mut Option<InteractiveTxSigningSession>, &mut ChannelContext<SP>);
```
Why is it important to have both returned in one call? Could two calls to `interactive_tx_signing_session_mut` and `context_mut` achieve the same?

jkczyz replied:
They both borrow mutable references, so the borrow checker will complain. Though maybe adding two lifetimes to the trait would get around it. Will give it a try.

optout21 commented on lightning/src/ln/channel.rs:
```diff
@@ -4026,21 +4049,15 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
-							"Failed to get signatures for new commitment_signed".to_owned(),
-							ClosureReason::HolderForceClosed { broadcasted_latest_txn: Some(false) },
-						)))
+					.ok()
```
What's the reasoning for discarding error info here? The signer may return an error.

jkczyz replied:
Signer returning an error indicates async signing. At least that's how the other async signing operates. Looks like the `EcdsaChannelSigner::sign_counterparty_commitment` docs weren't updated like `ChannelSigner::get_per_commitment_point` were. Let me do that in this PR.

@TheBlueMatt Wonder if we should change the interface to use an `Option` instead of a `Result`once `async_signing` configuration parameter is dropped?

optout21 commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8304,45 +8368,23 @@ where
+					match chan_phase_entry.get_mut() {
 						ChannelPhase::UnfundedOutboundV2(chan) => {
-							chan.funding_tx_constructed(&mut signing_session, &self.logger)
+							*chan.interactive_tx_signing_session_mut() = Some(signing_session);
```
Maybe a setter method would be more idiomatic in this case.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -9543,9 +9585,48 @@ where
+			None => {
+				let per_peer_state = self.per_peer_state.read().unwrap();
+				let mut channels = Vec::with_capacity(per_peer_state.len());
+				for (counterparty_node_id, peer_state_mutex) in per_peer_state.iter() {
```
`finish_tx_complete!` removes and re-inserts the entry, which you can't do while iterating. See TODO above that loop.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -8275,7 +8268,78 @@ where
 	}
+}
 
+macro_rules! finish_tx_complete {
```
Hmm... possibly. Maybe we should wait on #3423 before moving forward with this PR?

TheBlueMatt commented:
Dual funding is gonna slip yet again, setting to 0.4. Also needs rebase lol
